File: Grokking Algorithms/binary_search.cpp
#include <iostream>

int binarySearch(int* arr, size_t size, int num);

int main()
{
	int arr[]{1,3,5,7,9,11};
	size_t arrSize{sizeof(arr)/sizeof(int)};
	std::cout << binarySearch(arr, arrSize, 1) << '\n';
	std::cout << binarySearch(arr, arrSize, 3) << '\n';
	std::cout << binarySearch(arr, arrSize, 5) << '\n';
	std::cout << binarySearch(arr, arrSize, 7) << '\n';
	std::cout << binarySearch(arr, arrSize, 9) << '\n';
	std::cout << binarySearch(arr, arrSize, 11) << '\n';
	std::cout << binarySearch(arr, arrSize, 2) << '\n';

	return 0;
}

int binarySearch(int* arr, size_t size, int num)
{
	size_t low{0};
	size_t high{size - 1};

	while(low <= high)
	{
		size_t mid{(low + high) / 2};
		int guess{arr[mid]};
		if(guess == num)
			return mid;
		if(guess > num)
			high = mid - 1;
		else
			low = mid + 1;
	}

	return -1;
}

File: Grokking Algorithms/binary_search.py
def binary_search(list, item):
    low = 0
    high = len(list)-1

    while(low <= high):
        mid = low + high
        guess = list[mid]
        if(guess == item):
            return mid
        if(guess > item):
            high = mid - 1
        else:
            low = mid + 1
    return None

lst1 = [1,3,5,7,9,11]
print(binary_search(lst1, 1))
print(binary_search(lst1, 3))
print(binary_search(lst1, 5))
print(binary_search(lst1, 7))
print(binary_search(lst1, 9))
print(binary_search(lst1, 11))
print(binary_search(lst1, 222))

#1.1 log(2)128 = 7
#1.2 7+1
